cql3/query_processor.cc

```diff
@@ -1021,7 +1021,17 @@ query_processor::execute_schema_statement(const statements::schema_altering_stat
     stmt.global_req_id = mc.new_group0_state_id();
 
     auto [ce, warnings] = co_await stmt.prepare_schema_mutations(*this, state, options, mc);
-
+    // We are creating something.
+    if (!mc.empty()) {
+        // Internal queries don't trigger auto-grant. Statements which don't require
+        // auto-grant use default nop grant_permissions_to_creator implementation.
+        // Additional implicit assumption is that prepare_schema_mutations returns
+        // empty mutations vector when resource doesn't need to be created (e.g. already exists).
+        auto& client_state = state.get_client_state();
+        if (!client_state.is_internal()) {
+            co_await stmt.grant_permissions_to_creator(client_state, mc);
+        }
+    }
```

Comment:
A thought: Unless I'm missing something, my original patch plus this one would have fixed the permissions problem. The problem with my original patch was that the permission code used the is_schema_change() in the statement response to decide if to do this grant_permissions_to_creator. But this is exactly what you change here - after this change the call to grant_permissions_to_creator happens if we really created something - regarding of whether the "schema_change" event was sent in the reply or not.

If I understand correctly, this patch isn't strictly necessary for your own PR, right?

Reply:
Well, after reading the entire patch series it suddenly dawned on me - you actually switched your implementation to be the same one in my original patch :-) Which is why this change was needed.

Reply:
I think those changes are making code more readable

Reply:
> I don't have any complaints, but I'm puzzled why you ended changing your implementation after convincing me to accept it :-) Did other reviewers complain?

Yes Avi wanted this, I still thing it's wrong.

> Which is why this change was needed

It's the other way around, this change is needed, and on it's own it's a desired change as well. But it makes it possible to solve the problem by emiting event in a single place.

Reply:
> Yes Avi wanted this, I still thing it's wrong.

Then one of us should convince the other.

Reply:
And, why do we need both the change in query_processor and in the schema altering statement itself?

Reply:
I am convinced by 2 people out of 3 wanting this. Both ways will work so it's difficult to argue one way or another.